tsl/src/data_node.c

```diff
@@ -1415,6 +1428,240 @@ data_node_detach(PG_FUNCTION_ARGS)
+ * options or it is newly added. These are different operations on a foreign
+ * server.
+ *
+ * Any options that already exist are purged from the current_options list so
+ * that only the options not set or added remains. This list can be merged
+ * with the new options to produce the full list of options (new and old).
+ */
+static List *
+append_data_node_option(List *new_options, List **current_options, const char *name, Node *value)
+{
+	DefElem *elem;
+	ListCell *lc;
+	bool option_found = false;
+#if PG13_LT
+	ListCell *prev_lc = NULL;
+#endif
+
+	foreach (lc, *current_options)
+	{
+		elem = lfirst(lc);
+
+		if (strcmp(elem->defname, name) == 0)
+		{
+			option_found = true;
+			/* Remove the option which is replaced so that the remaining
+			 * options can be merged later into an updated list */
+#if PG13_GE
+			*current_options = list_delete_cell(*current_options, lc);
+#else
+			*current_options = list_delete_cell(*current_options, lc, prev_lc);
+#endif
+			break;
+		}
+#if PG13_LT
+		prev_lc = lc;
+#endif
+	}
+
+	elem = makeDefElemExtended(NULL,
+							   pstrdup(name),
+							   value,
+							   option_found ? DEFELEM_SET : DEFELEM_ADD,
+							   -1);
+	return lappend(new_options, elem);
+}
+
+/*
+ * Alter a data node.
+ *
+ * Change the configuration of a data node, including host, port, and
+ * database.
+ *
+ * Can also be used to mark a data node "unavailable", which ensures it is no
+ * longer used for reads as long as there are replica chunks on other data
+ * nodes to use for reads instead. If it is not possible to fail over all
+ * chunks, a warning will be raised.
+ */
+Datum
+data_node_alter(PG_FUNCTION_ARGS)
+{
+	const char *node_name = PG_ARGISNULL(0) ? NULL : NameStr(*PG_GETARG_NAME(0));
+	const char *host = PG_ARGISNULL(1) ? NULL : TextDatumGetCString(PG_GETARG_TEXT_P(1));
+	const char *database = PG_ARGISNULL(2) ? NULL : NameStr(*PG_GETARG_NAME(2));
+	int port = PG_ARGISNULL(3) ? -1 : PG_GETARG_INT32(3);
+	bool available_is_null = PG_ARGISNULL(4);
+	bool available = available_is_null ? true : PG_GETARG_BOOL(4);
+	ForeignServer *server = NULL;
+	List *current_options = NIL;
+	List *options = NIL;
+	TupleDesc tupdesc;
+	AlterForeignServerStmt alter_server_stmt = {
+		.type = T_AlterForeignServerStmt,
+		.servername = node_name ? pstrdup(node_name) : NULL,
+		.has_version = false,
+		.version = NULL,
+		.options = NIL,
+	};
+
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+
+	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("function returning record called in context "
+						"that cannot accept type record")));
+
+	tupdesc = BlessTupleDesc(tupdesc);
+
+	/* Check if a data node with the given name actually exists, or raise an error. */
+	server = data_node_get_foreign_server(node_name, ACL_NO_CHECK, false, false /* missing_ok */);
+
+	if (host == NULL && database == NULL && port == -1 && available_is_null)
+		PG_RETURN_DATUM(
```

Comment:
If no arguments supplied, it will return current state?

Reply:
yes, that's the intention.